File: sample/src/main/java/com/polidea/rxandroidble/sample/util/HexString.java
package com.polidea.rxandroidble.sample.util;

public class HexString {

    final protected static char[] hexArray = "0123456789ABCDEF".toCharArray();

    public static String bytesToHex(byte[] bytes) {
        char[] hexChars = new char[bytes.length * 2];

        for (int j = 0; j < bytes.length; j++) {
            int v = bytes[j] & 0xFF;
            hexChars[j * 2] = hexArray[v >>> 4];
            hexChars[j * 2 + 1] = hexArray[v & 0x0F];
        }

        return new String(hexChars);
    }

    public static byte[] hexToBytes(String hexRepresentation) {
        int len = hexRepresentation.length();
        byte[] data = new byte[len / 2];

        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(hexRepresentation.charAt(i), 16) << 4)
                    + Character.digit(hexRepresentation.charAt(i + 1), 16));
        }

        return data;
    }
}

File: sample/src/main/java/com/polidea/rxandroidble2/sample/util/HexString.kt
package com.polidea.rxandroidble2.sample.util

private val HEX_ARRAY = "0123456789ABCDEF".toCharArray()

internal fun ByteArray.bytesToHex(): String =
    CharArray(size * 2).let { hexChars ->
        for (j in indices) {
            val v = this[j].toInt() and 0xFF
            hexChars[j * 2] = HEX_ARRAY[v ushr 4]
            hexChars[j * 2 + 1] = HEX_ARRAY[v and 0x0F]
        }
        String(hexChars)
    }

internal fun String.hexToBytes(): ByteArray {
    if (length % 2 == 1) throw IllegalArgumentException("hexToBytes requires an even-length String parameter")

    return ByteArray(length / 2).also { data ->
        for (i in 0 until length step 2) {
            data[i / 2] = ((this[i].hexValue() shl 4) + this[i + 1].hexValue()).toByte()
        }
    }
}

private fun Char.hexValue(): Int = Character.digit(this, 16)
